pkg/image/apis/image/v1/helpers.go

```diff
@@ -0,0 +1,228 @@
+		if err := json.Unmarshal([]byte(manifest.History[0].DockerV1Compatibility), &v1Metadata); err != nil {
+			return err
+		}
+
+		image.DockerImageLayers = make([]ImageLayer, len(manifest.FSLayers))
+		for i, layer := range manifest.FSLayers {
+			image.DockerImageLayers[i].MediaType = schema1.MediaTypeManifestLayer
+			image.DockerImageLayers[i].Name = layer.DockerBlobSum
+		}
+		if len(manifest.History) == len(image.DockerImageLayers) {
+			// This code does not work for images converted from v2 to v1, since V1Compatibility does not
+			// contain size information in this case.
+			image.DockerImageLayers[0].LayerSize = v1Metadata.Size
+			var size = imageapi.DockerV1CompatibilityImageSize{}
+			for i, obj := range manifest.History[1:] {
+				size.Size = 0
+				if err := json.Unmarshal([]byte(obj.DockerV1Compatibility), &size); err != nil {
+					continue
+				}
+				image.DockerImageLayers[i+1].LayerSize = size.Size
+			}
+		}
+		// reverse order of the layers for v1 (lowest = 0, highest = i)
+		for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
+			image.DockerImageLayers[i], image.DockerImageLayers[j] = image.DockerImageLayers[j], image.DockerImageLayers[i]
+		}
+
+		dockerImage := &imageapi.DockerImage{}
+
+		dockerImage.ID = v1Metadata.ID
+		dockerImage.Parent = v1Metadata.Parent
+		dockerImage.Comment = v1Metadata.Comment
+		dockerImage.Created = v1Metadata.Created
+		dockerImage.Container = v1Metadata.Container
+		dockerImage.ContainerConfig = v1Metadata.ContainerConfig
+		dockerImage.DockerVersion = v1Metadata.DockerVersion
+		dockerImage.Author = v1Metadata.Author
+		dockerImage.Config = v1Metadata.Config
+		dockerImage.Architecture = v1Metadata.Architecture
+		if len(image.DockerImageLayers) > 0 {
+			size := int64(0)
+			layerSet := sets.NewString()
+			for _, layer := range image.DockerImageLayers {
+				if layerSet.Has(layer.Name) {
+					continue
+				}
+				layerSet.Insert(layer.Name)
+				size += layer.LayerSize
+			}
+			dockerImage.Size = size
+		} else {
+			dockerImage.Size = v1Metadata.Size
+		}
+
+		image.DockerImageMetadata.Object = dockerImage
+	case 2:
+		image.DockerImageManifestMediaType = schema2.MediaTypeManifest
+
+		if len(image.DockerImageConfig) == 0 {
+			return fmt.Errorf("dockerImageConfig must not be empty for manifest schema 2")
+		}
+		config := imageapi.DockerImageConfig{}
+		if err := json.Unmarshal([]byte(image.DockerImageConfig), &config); err != nil {
+			return fmt.Errorf("failed to parse dockerImageConfig: %v", err)
+		}
+
+		image.DockerImageLayers = make([]ImageLayer, len(manifest.Layers))
+		for i, layer := range manifest.Layers {
+			image.DockerImageLayers[i].Name = layer.Digest
+			image.DockerImageLayers[i].LayerSize = layer.Size
+			image.DockerImageLayers[i].MediaType = layer.MediaType
+		}
+		// reverse order of the layers for v1 (lowest = 0, highest = i)
+		for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
```

Comment:
I guess this is wrong. [The layer list is ordered starting from the base image (opposite order of schema1)](https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#image-manifest-field-descriptions)